### PR TITLE
 Check any recoloring retains the same parity of tiles A and B

### LINKED ISSUE #20: Any recoloring retains the same parity of tiles A and B
Generate all possible colorations of graphs of size 3*3.
Formed pairs, so that if one to couple (a,b) then one can get b from a by changing only the color of the central node of the graph.
Then check that the property, the parity of type A and B tiles is preserved after recoloring.

### PR DESCRIPTION
**Pull request by ThomArr**
Closes #20

### CODE REVIEW

File: lib/generator_graph.ml [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    [ [ 4; 1 ]; [ 1; 2 ] ];
+    [ [ 4; 1 ]; [ 1; 3 ] ];
+    [ [ 4; 1 ]; [ 3; 2 ] ];
+    [ [ 4; 3 ]; [ 1; 2 ] ];
+    [ [ 2; 3 ]; [ 4; 1 ] ];
+    [ [ 4; 2 ]; [ 1; 3 ] ];
+    [ [ 4; 3 ]; [ 3; 2 ] ];
+  ]
+
+let copy_matrix matrix =
```
Comment (GabinDDL) [inline lib/generator_graph.ml:21]: there is a function Array.copy, perhaps you can use it here

File: lib/generator_graph.ml [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+          [ [ h; i ]; [ b; c ] ];
+          [ [ c; a ]; [ d; f ] ];
+          [ [ f; d ]; [ i; g ] ];
+        ]
+    | _ -> []
+  in
+  if tiles_from_graph = [] then false
+  else
+    let acc =
+      List.fold_left
```
Comment (GabinDDL) [inline lib/generator_graph.ml:106]: I think you can create a function for this : number_tile_a_and_b or something like that

File: lib/generator_graph.ml [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        ((not parity_of_first_tile) || parity_of_second_tile)
+        && ((not parity_of_second_tile) || parity_of_first_tile)
+      in
+      for_all_pairs i (j + 1) (count_pair + 1)
+        (prop_check :: conserv_Parity_list)
+    else for_all_pairs i (j + 1) count_pair conserv_Parity_list
+  in
+  for_all_pairs 0 0 0 []
+
+let start () =
```
Comment (GabinDDL) [inline lib/generator_graph.ml:135]: a good thing to do is to add flags into main and move it into it

**Review by GabinDDL — COMMENTED** [review]
just some comments, except that it seems good for me, really good job, you're the best 🥇

File: lib/generator_graph.ml [hunk 4]
```diff
@@ -0,0 +1,149 @@
+let copy_matrix matrix =
+  Array.init (Array.length matrix) (fun i -> Array.copy matrix.(i))
+
+let check_neighbors_does_not_have_same_color grid height width current_row
```
Comment (GabinDDL) [inline lib/generator_graph.ml:4]: in this function, you don't check the modulo position ? ((0, 0 ) for (w - 1, 0) for exemple)

File: lib/generator_graph.ml [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+      else for_all_color (current_color + 1) acc
+    in
+    for_all_color 1 []
+
+let generate_all_colorings width height max_colors =
+  let grid = Array.make_matrix width height 0 in
+  generate_colorings grid width height 0 0 max_colors
+
+let check_graphs_differ_by_one_value_on_middle arr1 arr2 =
+  if Array.length arr1 <> Array.length arr2 then false
```
Comment (GabinDDL) [inline lib/generator_graph.ml:49]: perhaps retain the length values to avoid repeating the calculation

File: lib/generator_graph.ml [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    (fun acc e ->
+      if List.exists (fun x -> x = e) set_tilesA then acc + 1
+      else if List.exists (fun x -> x = e) set_tilesB then acc + 1
+      else acc)
+    0 set_tiles
+
+let is_parity_between_tiles_A_and_B graph set_tilesA set_tilesB =
+  let tiles_from_graph =
+    match graph with
+    | [| [| a; b; c |]; [| d; e; f |]; [| g; h; i |] |] ->
```
Comment (GabinDDL) [inline lib/generator_graph.ml:72]: there is toroidal tiles who miss in this : : [[g; h]; [a; b]] for example

**Comment by GabinDDL** [comment]
apart from my comments I have nothing to say, good job, you're the best!

File: lib/generator_graph.ml (hunk 6 above)
Comment (ThomArr) [reply lib/generator_graph.ml:72]: Not precisely the teacher specified that they were not

**Review by GabinDDL — APPROVED** [review]

**Pull request merged**